H.Ws/H.W_4.py

```diff
@@ -0,0 +1,236 @@
+)  # .dropna().reset_index()
+
+print(titanic_df)
+print(titanic_df.dtypes)
+print(titanic_df.columns)
+predictors = ["pclass", "sex", "age", "sibsp", "embarked", "class"]
+response = ["alive"]
+
+x = titanic_df[predictors]
+y = titanic_df[response]
+resp_test = None
+pred_test = None
+
+# for response
+for val in response:
+    if y[val].dtype == "bool":
+        resp_test = "boolean"
+    else:
+        resp_test = "continuous"
+
+# for predictors
+for val in predictors:  # x.columns
+    if (
+        x[val].dtype == "string"
+        or x[val].dtype == "object"
+        or x[val].dtype == "bool"
+        or x[val].dtype == "category"
+    ):
+        pred_test = "categorical"
+    else:
+        pred_test = "continuous"
+
+
+# predictor - cont: Age
+# predictor - cat: sex, alone
+# response - cat: survived, alive
+
+# build code for all 4 senarios
+
+# 1_cat response - cat predictor
+def cat_response_cat_predictor():
+    conf_matrix = confusion_matrix(x[val], y[val])
+
+    fig_no_relationship = go.Figure(
+        data=go.Heatmap(z=conf_matrix, zmin=0, zmax=conf_matrix.max())
+    )
+    fig_no_relationship.update_layout(
+        title="Categorical Predictor by Categorical Response (with relationship)",
+        xaxis_title="Response",
+        yaxis_title="Predictor",
+    )
+    fig_no_relationship.show()
+    pass
+
+
+if resp_test == "boolean" and pred_test == "categorical":
+    for i in x and y:
+        cat_response_cat_predictor()
+
+
+# 2_cat response - cont predictor
+def cat_resp_cont_predictor():
+    n = 200
+    # test matrix
+    # x = titanic_df['age'].to_numpy()
+    # x = x[~numpy.isnan(x)]
+    # y = titanic_df['fare'].to_numpy()
+    # y = y[~numpy.isnan(y)]
+    # z = [x, y]
+
+    group_labels = ["Response"]
+
+    # Create distribution plot with custom bin_size
+    fig_1 = ff.create_distplot([x[val]], group_labels, bin_size=0.2)
+    fig_1.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title="Predictor",
+        yaxis_title="Distribution",
+    )
+    fig_1.show()
+
+    fig_2 = go.Figure()
+    for y, curr_group in zip([x[val]], group_labels):
+        fig_2.add_trace(
+            go.Violin(
+                x=numpy.repeat(curr_group, n),
+                y=y,
+                name=curr_group,
+                box_visible=True,
+                meanline_visible=True,
+            )
+        )
+    fig_2.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title="Response",
+        yaxis_title="Predictor",
+    )
+    fig_2.show()
+    pass
+
+
+if resp_test == "boolean" and pred_test == "continuous":
+    for i in x:
+        cat_resp_cont_predictor()
+
+
+# 3_con response - cat predictor
+def cont_resp_cat_predictor():
+    n = 200
+    # Add histogram data
+    # a = titanic_df['sex']
+    # b = titanic_df['pclass']
+    # c = titanic_df['deck']
+    # z = [a, b, c]
+
+    group_labels = ["Group 1", "Group 2", "Group 3"]
```

Comment:
I am not sure if it would be a good idea to hardcode these values..if you test it against a new dataframe with different number of unique values of the categorical predictor, then it will fail..